bpo-45383: Inherit MetaClass from bases in FromSpec API

seberg opened the pull request:
This checks the bases of of a type created using the FromSpec
API to inherit the bases metaclasses.  The MetaClasses alloc
function will be called as is done in `tp_new` for classes
created in Python.

---

I have tested that both tests fail on the dev branch without the fix.  This doesn't explicitly test that using `tp_bases` also works (yet).  It also doesn't test the allocation size, the way it is written if the metaclass is set correctly (through allocation), then the allocation should also be correct.
(I have not done that, because that would require creating a MetaClass that extends the `PyHeaptType_Object`.  Certainly possible, but it seemed a bit much – although that is my secret agenda ;)).

<!-- issue-number: [bpo-45383](https://bugs.python.org/issue45383) -->
https://bugs.python.org/issue45383
<!-- /issue-number -->

---

- [x] Test that `__subclasses__` is set correctly
- [x] Call type_new_init_subclass()?
- [x] Explicitly test that using `tp_bases` also works
- [ ] ~~Check the allocation size?~~

seberg committed:
BUG: Inherit MetaClass from bases in FromSpec API

This checks the bases of of a type created using the FromSpec
API to inherit the bases metaclasses.  The MetaClasses alloc
function will be called as is done in `tp_new` for classes
created in Python.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

seberg commented on Modules/_testcapimodule.c:
```diff
@@ -1173,6 +1173,143 @@ test_get_type_name(PyObject *self, PyObject *Py_UNUSED(ignored))
+static PyType_Spec MinimalType_spec = {
+    "_testcapi.MinimalSpecType",
+    0,
+    0,
```
This fails the tests with assertions, but since I am unsure about the random choice of using `abc.ABC` and `ctypes.Array` for the tests (or the test layout).  I would be happy about feedback before fixing it (I don't use FromSpec much, so thought 0 might mean "inherit").

encukou replied:
No, you need to put the full size here. You'd use `sizeof(the struct for the object)` here, and that struct needs to include the superclass' struct at the start for subclassing to work.

But making -1 mean "inherit" would be a nice addition for cases where you aren't adding any new C-level state. I filed https://bugs.python.org/issue45525.

seberg replied:
Well, tried using only ctypes classes, which requires a bit less of a dance (although annoying includes, which I am a bit unsure about).

seberg replied:
Ok, the last commit was mostly useless :).  I can't access those symbols through C, since `ctypes` is its own compilation unit...  Could still create my own metaclasses, I guess.

encukou replied:
And I closed that issue – an option for "inherit" wouldn't be very useful generally.

>  Could still create my own metaclasses, I guess.

That would be best. Relying on details in other libraries tends to break after a few years, and metaclass/inheritance patterns are definitely such details.

encukou commented on Objects/typeobject.c:
```diff
@@ -3456,12 +3431,47 @@ PyType_FromModuleAndSpec(PyObject *module, PyType_Spec *spec, PyObject *bases)
         Py_INCREF(bases);
     }
 
+    /* NOTE: Missing API to replace `&PyType_Type` below, see bpo-15870 */
```
This is not a problem in CPython itself.

seberg replied:
Of course not, I just removed it, I guess `Missing API to customize` would be clearer, but it doesn't matter.  The bpo exists and it is trivial to find the right place.

encukou commented on Objects/typeobject.c:
```diff
@@ -3398,37 +3398,12 @@ PyType_FromModuleAndSpec(PyObject *module, PyType_Spec *spec, PyObject *bases)
         PyErr_SetString(PyExc_SystemError,
                         "Type spec does not define the name field.");
-        goto fail;
+        return NULL;
```
Instead of this, could you keep the `goto fail`s, but set `res = NULL` at the start and use [`Py_XDECREF`](https://docs.python.org/3/c-api/refcounting.html#c.Py_XDECREF) in the `fail` section?
That style makes it easy to add other references in the future. Right away you can add `bases` (setting it to NULL after it should no longer be XDECREF'd).

seberg replied:
Agreed, done (I thought that might be more "minimal" but it really was not).

encukou reviewed (commented):
I have a few early comments. I'll continue to review tomorrow.

seberg committed:
MAINT: Address review comments

First, use `goto fail` unconditionally, and second try to simplify
the tests by relying only on ctypes private (MetaClass using) API.

encukou committed:
Merge in the main branch

encukou committed:
Disallow non-type metaclasses

encukou committed:
Use custom metaclasses for the tests, rather than stuff from ctypes

encukou committed:
Make PyType_FromMetaclass refcounting easier to follow

- Use separate variables for bases (strong ref) and basesnin (borrowed)
- Split out bases tuple preparation
- Decref held references at the end, in both success and error cases

encukou committed:
Adjust documentation

seberg commented on Objects/typeobject.c:
```diff
@@ -3420,88 +3449,89 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
+        goto finally;
     }
-    res = (PyHeapTypeObject*)metatype->tp_alloc(metatype, nmembers);
+    if (metaclass->tp_new != PyType_Type.tp_new) {
```
I am now thinking that this check should probably be:
```
provided_metaclass->tp_new != metaclass->tp_new
```
That will (a bit more precise) allow creating an instance of a metaclass with a custom `tp_new`, so long you pass said metaclass explicitly.

EDIT: To be clear, that is if there is a base that is a metaclass instance itself!

encukou replied:
The problem I see is that we can't really *call* `tp_new` properly, since we're not getting the arguments to pass to it.
We could call it without arguments, but that sounds half-baked.

Do you have a use case for allowing custom `tp_new`?

seberg replied:
My thinking was that when a user passes `metaclass` explicitly, we can assume they know how to initialize it properly.

However, you are right.  I don't have an explicit use-case (since I can disable the custom `tp_new` in NumPy, I think), so it seems just as well to not anticipate this.  Especially since it could probably be relaxed later if a use-case arises.

encukou replied:
I added this caveat to the docs:

>   Note that this function does *not* fully match the behavior of  calling `type` or using the `class` statement.
   With user-provided base types or metaclasses, prefer  calling `type (or the metaclass)   over ``PyType_From*`` functions.
   Specifically:
>   * :py:meth:`~object.__new__` is not called on the new class
     (and it must be set to ``type.__new__``).
>   * :py:meth:`~object.__init__` is not called on the new class.
>   * :py:meth:`~object.__init_subclass__` is not called on any bases.
>   * :py:meth:`~object.__set_name__` is not called on new descriptors.

This *could* be solved by accepting `*kwargs` and calling the relevant bits, but I'm worried that there are subtler edge cases I missed. IMO it's best to leave this until a use case is found.

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3518,19 +3540,21 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
     type->tp_as_buffer = &res->as_buffer;
+
     /* Set tp_base and tp_bases */
+    type->tp_base = Py_NewRef(base);
```
```suggestion
    type->tp_base = (PyTypeObject *)Py_NewRef(base);
```

erlend-aasland replied:
Oops, leftovers from early review.

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3365,30 +3365,59 @@ static const PySlot_Offset pyslot_offsets[] = {
+    PyObject *modname = NULL;
+    PyTypeObject *type;
+    PyObject *bases = NULL;
+    PyObject *result = NULL;
```
```suggestion
```

erlend-aasland replied:
Sorry, leftovers from early review :)

encukou committed:
Fix type warning

encukou committed:
Remove unused includes from _testcapimodule.c

encukou committed:
Use an explicit role in the News entry

encukou committed:
Test that __subclasses__ is updated properly

encukou committed:
Document caveats, suggest that PyType_From* is not for user-provided classes

encukou committed:
Remove unused variable

encukou committed:
Move bases calculation before the allocation

encukou committed:
Fix compiler warning for Py_NewRef result

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3365,30 +3365,58 @@ static const PySlot_Offset pyslot_offsets[] = {
+ * types), return a tuple of types.
+ */
+inline static PyObject *
+get_bases_tuple(PyObject *bases_in, PyType_Spec *spec) {
```
Style nit:

```suggestion
get_bases_tuple(PyObject *bases_in, PyType_Spec *spec)
{
```

encukou replied:
PEP 7 strikes again! I still don't have this style in finger memory, alas.

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3365,30 +3365,58 @@ static const PySlot_Offset pyslot_offsets[] = {
-    PyObject *modname;
-    PyTypeObject *type, *base;
+    PyHeapTypeObject *res = NULL;
+    PyObject *modname = NULL;
```
We should move `modname` down to the `if` scope where it belongs. (Maybe save that for another PR.)

encukou replied:
Yeah, another PR with some clean-up is coming up :)

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3365,30 +3365,58 @@ static const PySlot_Offset pyslot_offsets[] = {
+    PyTypeObject *type;
+    PyObject *bases = NULL;
     int r;
+    int success = 0;
```
Why not just use `PyErr_Occurred` at the end?

encukou replied:
Good idea!

erlend-aasland commented on Objects/typeobject.c:
```diff
@@ -3417,17 +3445,61 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
+                     "type '%.100s' is not an acceptable base type",
+                     base->tp_name);
+        goto finally;
+    }
```
This should not be needed; `best_base()` already performs this check.

```suggestion
```

erlend-aasland replied:
Alternatively make it an assert?

erlend-aasland reviewed (commented):
Looks good! Some comments:

- Would it be wort it to use `_PyTuple_FromArray` to speed things up? I guess not, and if yes, it's out of scope of this PR anyway :)
- I'd prefer to only do test setup in `_testcapimodule.c`, and perform the actual tests in `Lib/test/test_capi.py`

encukou committed:
Use PyErr_Occurred rather than a variable

encukou committed:
Use PEP 7 style

encukou committed:
Replace duplicate Py_TPFLAGS_BASETYPE check with an assert

encukou reviewed (commented):
Where would you suggest `_PyTuple_FromArray`?

> I'd prefer to only do test setup in _testcapimodule.c, and perform the actual tests in Lib/test/test_capi.py

That would mean the test is split across two files, making it harder to read...
